### PR TITLE
OIDC - Adding HashiCorp Vault setup instructions

### PR DESCRIPTION
**Pull request by smaeda-ks**
### Why:

The current page is missing the detailed Vault setup instructions:
https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault

### What's being changed:

Adding more Vault setup instructions.

### Check off the following:

- [x] I have reviewed my changes in staging (look for "Automatically generated comment" and click **Modified** to view your latest changes).
- [x] For content changes, I have completed the [self-review checklist](https://github.com/github/docs/blob/main/contributing/self-review.md#self-review).

### Writer impact (This section is for GitHub staff members only):

- [ ] This pull request impacts the contribution experience
  - [ ] I have added the 'writer impact' label
  - [ ] I have added a description and/or a video demo of the changes below (e.g. a "before and after video")

### CODE REVIEW

**Comment by ramyaparimi** [comment]
@smaeda-ks Thanks so much for opening this PR. We don't accept changes made to this file `content/actions/deployment/security-hardening-your-deployments/**` in the open source repository. 

I am going to close this now 💛 Thanks so much for taking your time in improving the GitHub docs 💖

**Comment by smaeda-ks** [comment]
Hi @ramyaparimi, I'm trying to understand whether you will be picking up this PR in your private repo and syncing or you're simply denying this PR at all? Can you please clarify what will you be doing next?

**Comment by janiceilene** [comment]
Thanks so much for that clarification @smaeda-ks! We didn't realize you were a partner 💛 I'm reopening this, deleting all of the bot comments above 😅 , and we'll work with the writers to get a review.

**Comment by martin389** [comment]
Thanks for this PR! ✨ I'll work on getting this reviewed 👍

**Comment by elpoll** [comment]
That's weird

**Comment by elpoll** [comment]
I'm new

**Comment by smaeda-ks** [comment]
Thanks @martin389 ! Let me know if you have any questions or I should make any edits

**Comment by martin389** [comment]
👋 @smaeda-ks - sorry about the delay, I'll follow up with the reviewers.

**Comment by ethomson** [comment]
Hi @smaeda-ks - it looks like you and @ksatirli are both working on improving the Vault documentation. @ksatirli has opened https://github.com/github/docs/pull/16311

That's great! We're really excited to improve this documentation. But I'm hopeful that we can take the best of both of these guidances. It looks like you might both work at HashiCorp? If so, I'm hoping that you two can work together to provide a single set of guidance? 🙏 

Hopefully we can agree on one of these two PRs to move forward with, bring any improvements from the other into the "canonical" PR and then close the other? After that we'll have our OIDC team do a final review and merge the PR.

Thanks much!

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 1]
```diff
@@ -99,6 +99,14 @@ This example demonstrates how to use OIDC with the official action to request a
 
  {% data reusables.actions.oidc-permissions-token %}
 
+{% note %}
+
+**Note**:
+
+When the `permissions` key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access. Therefore, you may need to add other permissions, such as `content: read`. See [Automatic token authentication](/actions/security-guides/automatic-token-authentication) for more information.
```
Comment (ksatirli) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:106]:
```suggestion
When the `permissions` key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access. Therefore, you may need to add other permissions, such as `contents: read`. See [Automatic token authentication](/actions/security-guides/automatic-token-authentication) for more information.
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 2]
```diff
@@ -99,6 +99,14 @@ This example demonstrates how to use OIDC with the official action to request a
...
+
+{% endnote %}
+
 ### Requesting the access token
 
 The `hashicorp/vault-action` action receives a JWT from the {% data variables.product.prodname_dotcom %} OIDC provider, and then requests an access token from your HashiCorp Vault instance to retrieve secrets. For more information, see the HashiCorp Vault [documentation](https://github.com/hashicorp/vault-action).
```
Comment (ksatirli) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:112]:
```suggestion
The `hashicorp/vault-action` action receives a JWT from the {% data variables.product.prodname_dotcom %} OIDC provider, and then requests an access token from your HashiCorp Vault instance to retrieve secrets. For more information, see the HashiCorp Vault GitHub Action [documentation](https://github.com/hashicorp/vault-action).
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md (hunk 1 above)
Comment (smaeda-ks) [reply content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md]: ah, thanks for catching this! @ksatirli

**Comment by smaeda-ks** [comment]
Hi @ethomson and @martin389,

@ksatirli and I worked together and finalized this PR and this is now ready for your review. Sorry for taking longer to update, and we appreciate your help.

**Comment by ethomson** [comment]
Thanks @smaeda-ks! This looks good to me, I'll ask another team member for a set of :eyes: as well.

**Comment by smaeda-ks** [comment]
I have resolved the outstanding conflicts.

**Comment by smaeda-ks** [comment]
Hi team @ramyaparimi @janiceilene @martin389 @ethomson,

Can we get an update on this review? Thanks in advance!

**Comment by parthd24** [comment]
@N-Usha and I were hoping that the cloud specific OIDC steps are better off in HashiCorp’s documentation. We've re-pinged Alex Cahn on the HashiCorp Tech Alliances team on this just now

**Comment by N-Usha** [comment]
> As an end-user of the code that's shown here, I would expect to find it on GitHub Docs, as it provides configuration that is very specific to GitHub.

@ksatirli - Fair expectation. However we would prefer that GitHub docs focus on the core OIDC support in GitHub Actions  while the respective cloud’s documentation cover the OIDC setup for their cloud roles (config on Vault server in Hashi's case). For example, Hashi could have an article dedicated to GitHub OIDC similar to [this](https://learn.hashicorp.com/tutorials/vault/agent-aws?in=vault/auth-methods))
And then GitHub docs could cross link these docs to ensure that the discoverability exists from both entry points (Cloud, GitHub).  This IMHO would ensure better maintainability as the respective clouds take up the onus of maintaining their content. Hope that explains our rationale.

**Comment by jaystyles** [comment]
@dependabot merge

**Comment by jaystyles** [comment]
@dependabot close

**Comment by jaystyles** [comment]
@ dependabot merge

**Comment by jaystyles** [comment]
@ dependabot close

**Comment by ksatirli** [comment]
@N-Usha thank you for your patience - I realized that I didn't reply to you here.

I agree that GitHub Docs should focus on the core OIDC support and not have overly cloud-specific information. 

Our change was built on top of _already_ available documentation that is currently available at https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault

The proposed changes are not net-new, but rather expand upon what is already available, with the goal of making it easier for our shared end-users to use GitHub OIDC with HashiCorp Vault. Better security for everyone is a win for all of us.

Other than the change on [line 135](https://github.com/github/docs/pull/15980/files#diff-5249305c1759c919506abfa21e97080e2e795509b8c8014abf0674e33133e91fR135), _all_ changes will work on Vault OSS (which is how we wrote the document).

Does this help with setting the context of this change?

**Comment by N-Usha** [comment]
Thanks for the additional context @ksatirli. In the best interest of helping end users, we will just go ahead and merge this documentation. We'll work with our docs team to confirm if we could reopen this PR or the needed next steps.
However appreciate if you could also kindly update the Hashi documentation with an explicit article/section for guiding the Vault users targeting to integrate with GitHub OIDC provider and perhaps cross reference the current GitHub article? Thanks!

**Comment by ksatirli** [comment]
@N-Usha thanks for taking the time to consider this. 

I will work with our internal team to ensure we get our documentation updated and cross-linked. I see great benefit in making this process so incredibly simple that it will see much higher adoption.

Thanks, again!

**Comment by N-Usha** [comment]
@parthd24 for visibility. Thanks!

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 3]
```diff
@@ -49,39 +92,98 @@ To add OIDC integration to your workflows that allow them to access secrets in V
...
 
 ### Adding permissions settings
 
  {% data reusables.actions.oidc-permissions-token %}
 
+{% note %}
+
+**Note**:
+
+When the `permissions` key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access. Therefore, you may need to add other permissions, such as `contents: read`. See [Automatic token authentication](/actions/security-guides/automatic-token-authentication) for more information.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:107]:
```suggestion
When the `permissions` key is used, all unspecified permissions are set to _no access_, with the exception of the metadata scope, which always gets _read_ access. As a result, you may need to add other permissions, such as `contents: read`. See [Automatic token authentication](/actions/security-guides/automatic-token-authentication) for more information.
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 4]
```diff
@@ -32,10 +32,53 @@ This guide gives an overview of how to configure HashiCorp Vault to trust {% dat
 
 To use OIDC with HashiCorp Vault, you will need to add a trust configuration for the {% data variables.product.prodname_dotcom %} OIDC provider. For more information, see the HashiCorp Vault [documentation](https://www.vaultproject.io/docs/auth/jwt).
 
-Configure the vault to accept JSON Web Tokens (JWT) for authentication:
-- For the `oidc_discovery_url`, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}
-- For `bound_issuer`, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}
-- Ensure that `bound_subject` is correctly defined for your security requirements. For more information, see ["Configuring the OIDC trust with the cloud"](/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#configuring-the-oidc-trust-with-the-cloud) and [`hashicorp/vault-action`](https://github.com/hashicorp/vault-action).
+To configure your Vault server to accept JSON Web Tokens (JWT) for authentication:
+
+1. Enable the JWT auth method and write configuration.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:37]:
```suggestion
1. Enable the JWT `auth` method, and use `write` to apply the configuration to your Vault. 
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 5]
```diff
@@ -32,10 +32,53 @@ This guide gives an overview of how to configure HashiCorp Vault to trust {% dat
...
-- For the `oidc_discovery_url`, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}
-- For `bound_issuer`, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}
-- Ensure that `bound_subject` is correctly defined for your security requirements. For more information, see ["Configuring the OIDC trust with the cloud"](/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#configuring-the-oidc-trust-with-the-cloud) and [`hashicorp/vault-action`](https://github.com/hashicorp/vault-action).
+To configure your Vault server to accept JSON Web Tokens (JWT) for authentication:
+
+1. Enable the JWT auth method and write configuration.  
+  For `oidc_discovery_url` and `bound_issuer` parameters, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}. These parameters must be set so that the Vault server can verify the received JSON Web Tokens (JWT) upon authentication.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:38]:
```suggestion
  For `oidc_discovery_url` and `bound_issuer` parameters, use {% ifversion ghes %}`https://HOSTNAME/_services/token`{% else %}`https://token.actions.githubusercontent.com`{% endif %}. These parameters allow the Vault server to verify the received JSON Web Tokens (JWT) during the authentication process.
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 6]
```diff
@@ -32,10 +32,53 @@ This guide gives an overview of how to configure HashiCorp Vault to trust {% dat
...
+    ```sh{:copy}
+    vault auth enable jwt
+    ```
+    
+    ```sh{:copy}
+    vault write auth/jwt/config \
+      bound_issuer="{% ifversion ghes %}https://HOSTNAME/_services/token{% else %}https://token.actions.githubusercontent.com{% endif %}" \
+      oidc_discovery_url="{% ifversion ghes %}https://HOSTNAME/_services/token{% else %}https://token.actions.githubusercontent.com{% endif %}"
+    ```
+2. Configure policies to only grant access for certain paths in which your workflows will be retrieving secrets. For more advanced policies, see the HashiCorp Vault [Policies documentation](https://www.vaultproject.io/docs/concepts/policies).
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:49]:
```suggestion
2. Configure a policy that only grants access to the specific paths your workflows will use to retrieve secrets. For more advanced policies, see the HashiCorp Vault [Policies documentation](https://www.vaultproject.io/docs/concepts/policies).
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 7]
```diff
@@ -32,10 +32,53 @@ This guide gives an overview of how to configure HashiCorp Vault to trust {% dat
...
+      },
+      "policies": ["myproject-production"],
+      "ttl": "10m"
+    }
+    EOF
+    ```
+
+- `ttl` defines the validity of the resulting access token.
+- Ensure that the `bound_claims` parameter is defined for your security requirements, and has at least one condition. Optionally, you can also set the `bound_subject` as well as the `bound_audiences` parameter.
+- The `bound_claims` parameter contains a set of claims and their required values to check arbitrary claims in the received JWT payload. In the above example, the role will accept any incoming authentication requests from the `repo-name` repository owned by the `user-or-org-name` account.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:78]:
```suggestion
- To check arbitrary claims in the received JWT payload, the `bound_claims` parameter contains a set of claims and their required values. In the above example, the role will accept any incoming authentication requests from the `repo-name` repository owned by the `user-or-org-name` account.
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 8]
```diff
@@ -49,39 +92,98 @@ To add OIDC integration to your workflows that allow them to access secrets in V
...
+      - name: Use secret from Vault
+        run: |
+          # This step has access to the secret retrieved above; see hashicorp/vault-action for more details.
+```
+
+{% note %}
+
+**Note**:
+
+- If your Vault server is not accessible from the public network, you may use the [self-hosted runners](/actions/hosting-your-own-runners/about-self-hosted-runners).
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:148]:
```suggestion
- If your Vault server is not accessible from the public network, consider using a self-hosted runner. For more information, see  "[About self-hosted runners](/actions/hosting-your-own-runners/about-self-hosted-runners)."
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 9]
```diff
@@ -49,39 +92,98 @@ To add OIDC integration to your workflows that allow them to access secrets in V
...
+- `<Vault Namespace>` must be set for a Vault Enterprise (including HCP Vault) deployment. See [Vault namespace](https://www.vaultproject.io/docs/enterprise/namespaces) for more information.
+
+{% endnote %}
+
+### Revoking the access token
+
+By default, the Vault server will automatically revoke access tokens when their TTL is expired, so you don't have to worry about revoking the access tokens.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:155]:
```suggestion
By default, the Vault server will automatically revoke access tokens when their TTL is expired, so you don't have manually revoke access tokens.
```

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 10]
```diff
@@ -49,39 +92,98 @@ To add OIDC integration to your workflows that allow them to access secrets in V
...
+
+But if you want to revoke access tokens immediately soon after your job is completed/failed for your security requirements, you can manually revoke the issued token using the [Vault API](https://www.vaultproject.io/api/auth/token#revoke-a-token-self).
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:157]:
```suggestion
However, if you do want to revoke access tokens immediately after your job has completed or failed, you can manually revoke the issued token using the [Vault API](https://www.vaultproject.io/api/auth/token#revoke-a-token-self).
```

**Review by martin389 — COMMENTED** [review]
👋 Added some suggestions to match the GitHub docs conventions

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md [hunk 11]
```diff
@@ -49,39 +92,98 @@ To add OIDC integration to your workflows that allow them to access secrets in V
...
+        run: |
+          # This step has access to the secret retrieved above; see hashicorp/vault-action for more details.
+```
+
+{% note %}
+
+**Note**:
+
+- If your Vault server is not accessible from the public network, you may use the [self-hosted runners](/actions/hosting-your-own-runners/about-self-hosted-runners).
+- `<Vault Namespace>` must be set for a Vault Enterprise (including HCP Vault) deployment. See [Vault namespace](https://www.vaultproject.io/docs/enterprise/namespaces) for more information.
```
Comment (martin389) [inline content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:149]:
```suggestion
- `<Vault Namespace>` must be set for a Vault Enterprise (including HCP Vault) deployment. For more information, see [Vault namespace](https://www.vaultproject.io/docs/enterprise/namespaces).
```

**Comment by martin389** [comment]
👋 Hi @smaeda-ks and @ksatirli -- I've added some suggestions that apply the GitHub docs conventions. Let me know if you're happy with these suggestions, and I'll be able to look at options for getting this PR merged.

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md (hunk 8 above)
Comment (ksatirli) [reply content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:148]: Much clearer, thanks!

File: content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md (hunk 9 above)
Comment (ksatirli) [reply content/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md:155]: Better, indeed!

**Comment by smaeda-ks** [comment]
@martin389 Thank you, I've merged those suggestions.

**Comment by martin389** [comment]
👋 Status update:

This PR changes files in the `content/actions/deployment/security-hardening-your-deployments/` directory, however, these files cannot be changed by PRs in this repo (see the [automated message](https://github.com/github/docs/pull/15980#issuecomment-1198706680) above). So in order to ship these changes, I'll recreate this PR in an internal repository, which will let us amend these files.

**Comment by ksatirli** [comment]
Thank you @martin389!

**Comment by 1996kaw** [comment]
> ## Automatically generated comment ℹ️
> **This comment is automatically generated and will be overwritten every time changes are committed to this branch.**
> 
> The table contains an overview of files in the `content` directory that have been changed in this pull request. It's provided to make it easy to review your changes on the staging site. Please note that changes to the `data` directory will not show up in this table.
> 
> ### Content directory changes
> _You may find it useful to copy this table into the pull request summary. There you can edit it to share links to important articles or changes and to give a high-level overview of how the changes in your pull request support the overall goals of the pull request._
> 
> **Source**	**Preview**	**Production**	**What Changed**
> [`actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault.md`](https://github.com/github/docs/blob/2b48589750fa13f35a8c7045d1af50e91f87c970/content%2Factions%2Fdeployment%2Fsecurity-hardening-your-deployments%2Fconfiguring-openid-connect-in-hashicorp-vault.md)	[fpt](https://docs-15980-94e1f2.preview.ghdocs.com/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)
> [ghec](https://docs-15980-94e1f2.preview.ghdocs.com/enterprise-cloud@latest/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)
> ghes@ [3.6](https://docs-15980-94e1f2.preview.ghdocs.com/enterprise-server@3.6/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault) [3.5](https://docs-15980-94e1f2.preview.ghdocs.com/enterprise-server@3.5/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)
> [fpt](https://docs.github.com/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)
> [ghec](https://docs.github.com/enterprise-cloud@latest/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)
> ghes@ [3.6](https://docs.github.com/enterprise-server@3.6/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault) [3.5](https://docs.github.com/enterprise-server@3.5/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault)

**Comment by martin389** [comment]
I've now recreated this PR internally, and will close this one while we work through the shipping processes.  I'll be away for a few days, so I've asked the team to continue work on this, depending on their availability. Otherwise I'll be able to resume this when I'm back 👍

**Comment by martin389** [comment]
Status update: I'm continuing work to get these changes shipped, and will comment here when they're live

**Comment by ksatirli** [comment]
Thank you for the update @martin389!

**Comment by martin389** [comment]
The changes are now live: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-hashicorp-vault 🎊

**Comment by ksatirli** [comment]
Thank you for your work on this @martin389 - very excited this made it into the docs!

**Pull request closed without merging**